runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkInputDataProcessor.java

```diff
@@ -0,0 +1,371 @@
+        throw re;
+      }
+    }
+
+    private void fireTimer(TimerInternals.TimerData timer) {
+      StateNamespace namespace = timer.getNamespace();
+      checkArgument(namespace instanceof StateNamespaces.WindowNamespace);
+      BoundedWindow window = ((StateNamespaces.WindowNamespace) namespace).getWindow();
+      ctx.getDoFnRunner()
+          .onTimer(
+              timer.getTimerId(),
+              timer.getTimerFamilyId(),
+              ctx.getKey(),
+              window,
+              timer.getTimestamp(),
+              timer.getOutputTimestamp(),
+              timer.getDomain());
+    }
+  }
+}
+
+class BoundedSparkInputDataProcessor<FnInputT, FnOutputT>
+    implements SparkInputDataProcessor<FnInputT, FnOutputT, Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+  private final BoundedDoFnOutputManager outputManager = new BoundedDoFnOutputManager();
+
+  @Override
+  public OutputManager getOutputManager() {
+    return outputManager;
+  }
+
+  @Override
+  public <K> Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> process(
+      Iterator<WindowedValue<FnInputT>> input, SparkProcessContext<K, FnInputT, FnOutputT> ctx) {
+    return new BoundedInOutIterator<>(input, ctx);
+  }
+
+  /**
+   * Output manager which can hold limited number of output elements. If capacity is reached, then
+   * attempt to output more elements will block until some elements are consumed.
+   */
+  private static class BoundedDoFnOutputManager
+      implements OutputManager, Iterable<Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+    private final LinkedBlockingQueue<Tuple2<TupleTag<?>, WindowedValue<?>>> queue =
+        new LinkedBlockingQueue<>(500);
+    private volatile boolean stopped = false;
+
+    public void stop() {
+      stopped = true;
+    }
+
+    @Override
+    public Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> iterator() {
+      return new Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>>() {
+
+        private @Nullable Tuple2<TupleTag<?>, WindowedValue<?>> value = null;
+
+        @Override
+        public boolean hasNext() {
+          // expect elements appearing in queue until stop() is invoked.
+          // after that, no more inputs can arrive, so just drain the queue
+          while (true) {
+            try {
+              value = queue.poll(1000, TimeUnit.MILLISECONDS);
+              if (value != null) {
+                return true;
+              }
+              if (stopped) {
+                return false;
+              }
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+              throw new IllegalStateException(e);
+            }
+          }
+        }
+
+        @Override
+        @SuppressWarnings({"nullness"})
+        public Tuple2<TupleTag<?>, WindowedValue<?>> next() {
+          return value;
+        }
+      };
+    }
+
+    @Override
+    public synchronized <T> void output(TupleTag<T> tag, WindowedValue<T> output) {
```

Comment:
```suggestion
    public <T> void output(TupleTag<T> tag, WindowedValue<T> output) {
```
the blocking queue is thread safe, no need to synchronize here

Reply:
Fixed